pkg/pilot/genericpilot/controller_test.go

```diff
@@ -0,0 +1,268 @@
+		return
+	}
+
+	// the fake scheduler above remove elements from this slice, so if any are
+	// left here then an item that should have been scheduled was not
+	if len(expectedScheduledItems) > 0 {
+		t.Errorf("Expected scheduled items %#v to be scheduled but not found in scheduled queue", expectedScheduledItems)
+		return
+	}
+
+	// get a list of actions performed on the navigator client
+	navActions := navClient.Actions()
+	for i, action := range navActions {
+		// check for unexpected actions
+		if len(f.navActions) < i+1 {
+			t.Errorf("%d unexpected actions: %+v", len(navActions)-len(f.navActions), navActions[i:])
+			break
+		}
+
+		expectedAction := f.navActions[i]
+		if !(expectedAction.Matches(action.GetVerb(), action.GetResource().Resource) && action.GetSubresource() == expectedAction.GetSubresource()) {
+			t.Errorf("Expected\n\t%#v\ngot\n\t%#v", expectedAction, action)
+			continue
+		}
+	}
+
+	if len(f.navActions) > len(navActions) {
+		t.Errorf("%d additional expected actions:%+v", len(f.navActions)-len(navActions), f.navActions[len(navActions):])
+	}
+}
+
+func newUpdatePilotStatusAction(p *v1alpha1.Pilot) core.Action {
+	a := core.NewUpdateAction(schema.GroupVersionResource{
+		Group:    navigator.GroupName,
+		Resource: "pilots",
+		Version:  "v1alpha1",
+	}, p.Namespace, p)
+	a.Subresource = "status"
+	return a
+}
+
+// returns a 'default' process adapter that always returns true for Running()
+func newDefaultProcessAdapter() process.Interface {
+	return &process.FakeAdapter{
+		RunningFunc: func() bool { return true },
+		StringFunc:  func() string { return "ps-adapter" },
+	}
+}
+
+func TestSync(t *testing.T) {
+	tests := []*fixture{
+		{
+			name:                   "successful run of Sync on 'this' pilot should return no error and update status",
+			key:                    "namespace/name",
+			err:                    false,
+			thisPilotName:          "name",
+			thisPilotNamespace:     "namespace",
+			expectedScheduledItems: []interface{}{"namespace/name"},
+			pilotLister:            []*v1alpha1.Pilot{newDummyPilot("name", "namespace")},
+			kubeObjects:            []runtime.Object{newDummyPilot("name", "namespace")},
+			navActions:             []core.Action{newUpdatePilotStatusAction(newDummyPilot("name", "namespace"))},
+			process:                newDefaultProcessAdapter(),
+		},
+		{
+			name:                   "should not Start a subprocess when syncing a non-this pilot",
+			key:                    "namespace/not-this-pilot",
+			err:                    false,
+			thisPilotName:          "name",
+			thisPilotNamespace:     "namespace",
+			expectedScheduledItems: []interface{}{"namespace/not-this-pilot"},
+			cachedThisPilot:        newDummyPilot("name", "namespace"),
+			pilotLister:            []*v1alpha1.Pilot{newDummyPilot("not-this-pilot", "namespace")},
+			kubeObjects:            []runtime.Object{newDummyPilot("not-this-pilot", "namespace")},
```

Comment:
I see that we don't supply a process in this test, but I don't understand how this is enough to demonstrate that a subprocess is *not* started.

Reply:
👍 my thinking was that the test will panic if process.{Start,Stop,Status etc..} is called, hence failing the test.

We could/should definitely extend this to actually just call `t.Fail` upon any calls to the processes methods (e.g. create a `NewFailingFakeProcess` function or something)